### PR TITLE
[Dynamic Fields] Replace Child Object docs with Dynamic Field Docs

### PR DESCRIPTION
**Pull request by amnn**
To complement the release of dynamic fields and an announcement to the community that child objects are going away.

Some changes in approach compared to the previous docs:

- I have not used snippets from the `object_owner` example, because they are more tailored to how child objects worked than how dynamic fields worked.
- ...which also means I have not included examples of interacting with these APIs using the CLI.

If these are still desirable for the docs, I would suggest doing this at a later stage, once dynamic fields have been fully released.

## Test Plan

👀

### CODE REVIEW

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+The first function accepts a mutable reference to the `Child` object directly, and can be called with `Child` objects that haven't been added as fields to `Parent` objects. Its body is empty since what we care about here is not how it is mutated, but whether the function can be called at all.
+
+The second functions accepts a mutable reference to the `Parent` object and accesses its dynamic field using `borrow_mut`, to pass to `mutate_child`. This can only be called on `Parent` objects that have a `b"child"` field defined. A `Child` object that has been added to a `Parent` *must* be accessed via its dynamic field, so it can only by mutated using `mutate_child_via_parent`, not `mutate_child`, even if its ID is known.
+
+> :warning: A transaction that attempts to borrow a field that does not exist will abort.
+
+> :warning: The `Value` type passed to `borrow` and `borrow_mut` must match the type of the stored field, or the transaction will abort.
+
+> :warning: A transaction that attempts to use a dynamic object field value as an input (by value or by reference), will fail to verify and therefore to run.
```
Comment (amnn) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:119]: @tnowacki -- want to confirm with you that this property is true (I'm recalling this property from reviewing the code).

Comment (tnowacki) [reply doc/src/build/programming-with-objects/ch5-dynamic-fields.md:119]: I would maybe reword it a bit. 
"Dynamic object field values _must_ be accessed through this API. A transaction that attempts to use those objects as inputs (by value or by reference) will be marked as having invalid inputs and will be rejected"

Bit of a run-on there at the end, but the point I'm trying to get is that I don't think you will even be able to generate a certificate for the transaction.

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
+---
+title: Chapter 5 - Dynamic Fields
+---
+
+In previous chapters, we walked through various ways to use object fields to store primitive data and other objects (wrapping), but there are a few limitations to this approach:
+
+1. Object's have a finite set of fields keyed by identifiers that are fixed when its module is published.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:7]: Maybe add
(i.e. limited to the fields in the  `struct` declration)

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+2. An object can become very large if it wraps several other objects. Larger objects can lead to higher gas fees in transactions. In addition, there is an upper bound on object size.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:8]: @oxade, do we have the size limit in place yet? I know there was some back and forth on that.

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+3. As we will see in future chapters, there will be use cases where we need to store a collection of objects of heterogeneous types. Since the Move `vector` type must be templated on one single type `T`, it is not suitable for this.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:9]:
```suggestion
3. As we will see in future chapters, there will be use cases where we need to store a collection of objects of heterogeneous types. Since the Move `vector` type must be instantiated with one single type `T`, it is not suitable for this.
```

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+Fortunately, Sui provides *dynamic fields* with arbitrary names (not just identifiers), added and removed on-the-fly (not fixed at publish), which only affect gas when they are accessed, and can store heterogeneous values. This chapter introduces the libraries for interacting with this kind of field.
+
+### Fields vs Object Fields
+
+There are two flavors of dynamic field -- "fields" and "object fields" -- which differ based on how their values are stored:
+
+- **Object field** values *must* be objects themselves (have the `key` ability, and `id: UID` as the first field) and will still appear under their ID to external tools (explorers, wallets, etc) accessing storage even when stored in a dynamic object field.
+- **Field** values do not need to be objects. An object stored in a (non-object) field will not be accessible via its ID while it is stored in a dynamic field.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:18]: nit, flip there order here to be consistent. I know it will require some rewording, but it is strange to go "Fields vs Object Fields" and then "fields and object fields", but then list "object field" first.

Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:18]:
```suggestion
- **Object field** values *must* be objects themselves (have the `key` ability, and `id: UID` as the first field) and will still appear under their ID to external tools (explorers, wallets, etc) accessing storage even when stored in a dynamic object field.
- **Field** values do not need to be objects. An object stored in a (non-object) field will be considered wrapped and will not be accessible via its ID while it is stored in a dynamic field.
```

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+The interfaces for interacting with these fields can be found in the [`dynamic_object_field`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/dynamic_object_field.move) and [`dynamic_field`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/dynamic_field.move) libraries respectively.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:20]: nit, flip the order again

Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:20]:
```suggestion
The modules for interacting with these fields can be found in the [`dynamic_object_field`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/dynamic_object_field.move) and [`dynamic_field`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/dynamic_field.move) libraries respectively.
```

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+### Adding Dynamic Fields
+
+Dynamic fields are added with the following APIs:
+
+```rust
+use sui::dynamic_field as field;
+use sui::dynamic_object_field as ofield;
+
+public fun field::add<Name: copy + drop + store, Value: store>(
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:34]: I get what's going on here... but still a bit strange as it isn't valid Move syntax. And might be confusing for someone without a programming language background where such a syntax makes sense. 

Maybe
```
module sui::dynamic_field {
...
public fun add<Name: copy + drop + store, Value: store>(
  object: &mut UID,
  name: Name,
  value: Value,
)
}
```
And similarly for `dynamic_object_field`

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+struct Child has key, store {
+    id: UID,
+}
+```
+
+Now, we can define an API to add a `Child` object as a dynamic field of a `Parent` object:
+
+```rust
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:63]: If the suggestion above is taken, please insert the `use` here

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+public entry fun add_child(parent: &mut Parent, child: Child) {
+    ofield::add(&mut parent.id, b"child", child);
+}
+```
+
+This function takes the `Child` object by value, and makes it a dynamic field of `parent` with name `b"child"` (a byte string). At the end of the `add_child` call, we have the following ownership relationship:
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:69]:
```suggestion
This function takes the `Child` object by value, and makes it a dynamic field of `parent` with name `b"child"` (a byte string of type `vector<u8>`). At the end of the `add_child` call, we have the following ownership relationship:
```

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+> :warning: It is an error to overwrite a field (attempt to add a field with the same Name type and value as one that is already defined), and a transaction that does this will abort.  Fields can be modified in-place by borrowing them mutably and can be overwritten safely (e.g. to change its value type) by removing the old value first (see below for details).
+
+### Accessing Dynamic Fields
+
+Dynamic fields can be accessed by reference using the following APIs:
+
+```rust
+use sui::dynamic_field as field;
+
+public fun borrow<Name: copy + drop + store, Value: store>(
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:83]: same `module sui::dynamic_field` suggestion as bove

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+public fun borrow_mut<Name: copy + drop + store, Value: store>(
+    object: &mut UID,
+    name: Name,
+): &mut Value
+```
+
+Where `object` is the UID of the object the field is defined on and `name` is the field's name.
+
+> :bulb: `sui::dynamic_object_field` has equivalent functions for object fields.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:96]:
```suggestion
> :bulb: `sui::dynamic_object_field` has equivalent functions for object fields, but with the added constraint `Value: key + store`
```

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+Let's look at how to use these APIs with the `Parent` and `Child` types defined earlier:
+
+```rust
+public entry fun mutate_child(_child: &mut Child) {}
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:101]: Amusingly... this won't actually result in a mutation in storage :p Unless you call `write_ref`, the global value won't be dirtied and it will be invisible to the object runtime (at least as it is currently written). 
Might want to just bump a counter or something?

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+> :warning: The `Value` type passed to `borrow` and `borrow_mut` must match the type of the stored field, or the transaction will abort.
+
+> :warning: A transaction that attempts to use a dynamic object field value as an input (by value or by reference), will fail to verify and therefore to run.
+
+### Removing a Dynamic Field
+
+Similar to "unwrapping" an object held in a regular field, a dynamic field can be removed, exposing its value:
+
+```
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:125]:
```suggestion
```rust
```

Comment (tnowacki) [reply doc/src/build/programming-with-objects/ch5-dynamic-fields.md:125]: And similarly below

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+use sui::dynamic_field as field;
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:126]: same comment about `module sui::dynamic_field {`

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md (hunk 3 above)
Comment (amnn) [reply doc/src/build/programming-with-objects/ch5-dynamic-fields.md:8]: IIRC there is not a direct object limit, but there is an indirect limit via gas (i.e. there is a gas limit, and therefore a max object size).

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md (hunk 5 above)
Comment (amnn) [reply doc/src/build/programming-with-objects/ch5-dynamic-fields.md:26]: Yeah, you're right... for whatever reason, it was easier for me to explain object fields first, but I seem to have gotten over that hang-up in the second draft!

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 15]
```diff
@@ -10,6 +10,14 @@ In previous chapters, we walked through various ways to use object fields to sto
 
 Fortunately, Sui provides *dynamic fields* with arbitrary names (not just identifiers), added and removed on-the-fly (not fixed at publish), which only affect gas when they are accessed, and can store heterogeneous values. This chapter introduces the libraries for interacting with this kind of field.
 
+### Current Limitations
+
+There are some aspects of dynamic fields that are not yet behaving as designed in this early release. We are actively working on these areas, but watch out for:
+
+- `remove` for dynamic fields not giving a full storage refund.
+- The lack of an `exists_` API to check whether a field with a particular name is already defined on an object.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:18]: It exists for `dynamic_object_field`, just not `dynamic_object`. For `dynamic_object`, we only have `exists_with_type`, which returns true iff it has a field with that name and the value has that type.

We probably also want `exists_` which would return true iff it has a field with that name (regardless of the dynamic value type)

File: doc/src/build/programming-with-objects/ch6-collections.md [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+public fun new<K: copy + drop + store, V: store>(
+    ctx: &mut TxContext,
+): Table<K, V>;
+
+}
+```
+
+`Table<K, V>` is a *homogeneous* map, meaning that all its keys have the same type as each other (`K`), and all its values have the same type as each other as well (`V`). It is created with `sui::table::new`, which requires access to a `&mut TxContext` because `Table`s are objects themselves, which can be transfered, shared, wrapped, or unwrapped, just like any other object.
+
+> :bulb: As usual, there is an equivalent `sui::object_table::ObjectTable` with an additional `V: key + store` constraint.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch6-collections.md:34]:
```suggestion
> :bulb: As with `dynamic_field` and `dynamic_object_field`, there is an equivalent `sui::object_table::ObjectTable` with an additional `V: key + store` constraint.
```

File: doc/src/build/programming-with-objects/ch6-collections.md [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+struct Bag has key, store { /* ... */ }
+
+public fun new(ctx: &mut TxContext): Bag;
+
+}
+```
+
+`Bag` is a *heterogeneous* map, so it can hold key-value pairs of arbitrary types (they don't need to match each other). Note that the `Bag` type does not have any type parameters for this reason. Like `Table`, `Bag` is also an object, so creating one with `sui::bag::new` requires supplying a `&mut TxContext` to generate an ID.
+
+> :bulb: See `sui::bag::ObjectBag` for the object-preserving version of `Bag`.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch6-collections.md:50]: I like this wording a lot. Maybe replicate it for the `table` bulb above?

File: doc/src/build/programming-with-objects/ch6-collections.md [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+module sui::table {
+
+public fun drop<K: copy + drop + store, V: drop + store>(
+    table: Table<K, V>,
+);
+
+}
+```
+
+that can only be called for tables where the both the key *and* value type have `drop` which allows it to delete tables whether they are empty or not.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch6-collections.md:169]: I find the wording here just a bit confusing, as the only new bit is the value part. The `K` type always needs `drop`

Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch6-collections.md:169]:
```suggestion
that can only be called for tables where the value type also has `drop`, which allows it to delete tables whether they are empty or not.
```

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 19]
```diff
@@ -28,20 +33,27 @@ Unlike an object's regular fields whose names must be Move identifiers, dynamic
 Dynamic fields are added with the following APIs:
 
 ```rust
-use sui::dynamic_field as field;
-use sui::dynamic_object_field as ofield;
+module sui::dynamic_field {
 
-public fun field::add<Name: copy + drop + store, Value: store>(
+public fun sui::dynamic_field::add<Name: copy + drop + store, Value: store>(
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:38]:
```suggestion
public fun add<Name: copy + drop + store, Value: store>(
```
we have the module, so no qualifier here

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 20]
```diff
@@ -28,20 +33,27 @@ Unlike an object's regular fields whose names must be Move identifiers, dynamic
...
+);
 
-public fun ofield::add<Name: copy + drop + store, Value: key + store>(
+}
+```
+
+```rust
+module sui::dynamic_object_field {
+
+public fun sui::dynamic_object_field::add<Name: copy + drop + store, Value: key + store>(
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:50]:
```suggestion
public fun add<Name: copy + drop + store, Value: key + store>(
```

File: doc/src/build/programming-with-objects/ch5-dynamic-fields.md [hunk 21]
```diff
@@ -161,6 +188,6 @@ public entry fun reclaim_child(parent: &mut Parent, ctx: &mut TxContext) {
 
 ### Deleting an Object with Dynamic Fields
 
-It is possible to delete an object that has dynamic fields still defined on it. Because field values can only be accessed via the dynamic field's associated object and field name, deleting an object that has dynamic fields still defined on it renders them all inaccessible to future transactions.
+It is possible to delete an object that has dynamic fields still defined on it. Because field values can only be accessed via the dynamic field's associated object and field name, deleting an object that has dynamic fields still defined on it renders them all inaccessible to future transactions. This is true regardless of whether the field's key or value has the `drop` ability.
```
Comment (tnowacki) [inline doc/src/build/programming-with-objects/ch5-dynamic-fields.md:191]: 👍

**Review by randall-Mysten — APPROVED** [review]
